security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -247,27 +256,37 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
 
 	// When there are existing root certificates, or private key and certificate under
 	// a well known path, they are used in the SDS response.
-	// In the current implementation, the file update events are not handled and if
-	// the files are updated, a user may restart Envoy to pick up the updated files.
-	// TODO (lei-tang): if updating files are supported, add a file watcher for
-	// the files under the well known path.
 	sdsFromFile := false
 	var err error
+	var ns *model.SecretItem
 
-	if connKey.ResourceName == RootCertReqResourceName && sc.rootCertificateExist(sc.existingRootCertFile) {
+	switch {
+	// Default root certificate.
+	case connKey.ResourceName == RootCertReqResourceName && sc.rootCertificateExist(sc.existingRootCertFile):
 		sdsFromFile = true
 		ns, err = sc.generateRootCertFromExistingFile(sc.existingRootCertFile, token, connKey)
-	} else if cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName); ok && cfg.CaCertificatePath != "" {
-		// Based on the resource name, we need to read this secret from a file encoded in the resource name
-		sdsFromFile = true
-		ns, err = sc.generateRootCertFromExistingFile(cfg.CaCertificatePath, token, connKey)
-	} else if connKey.ResourceName == WorkloadKeyCertResourceName && sc.keyCertificateExist(sc.existingCertChainFile, sc.existingKeyFile) {
+		sc.addFileWatcher(sc.existingRootCertFile, connKey)
+	// Based on the resource name, we need to read this secret from a file encoded in the resource name.
+	case connKey.ResourceName == WorkloadKeyCertResourceName && sc.keyCertificateExist(sc.existingCertChainFile, sc.existingKeyFile):
 		sdsFromFile = true
 		ns, err = sc.generateKeyCertFromExistingFiles(sc.existingCertChainFile, sc.existingKeyFile, token, connKey)
-	} else if cfg, ok := pilotmodel.SdsCertificateConfigFromResourceName(connKey.ResourceName); ok && cfg.CertificatePath != "" && cfg.PrivateKeyPath != "" {
-		// Based on the resource name, we need to read this secret from a file encoded in the resource name
-		sdsFromFile = true
-		ns, err = sc.generateKeyCertFromExistingFiles(cfg.CertificatePath, cfg.PrivateKeyPath, token, connKey)
+		// It is sufficient if we add one of cert or key file for watching.
```

Comment:
is there ever a case where one changes and not the other? is there harm in having both?

Reply:
I do not think so - but no harm in adding. I will add them